### PR TITLE
Feat: vela def gen-api command

### LINKED ISSUE #3606: Problem of converting batch applications
**Is your feature request related to a problem? Please describe.**
image
如图我想使用velacore通过api调用完成这类转换过程，但是velacore提供的create函数和我直接写一份yaml提交并无区别，大多数的参数都是map[string]interface{},我需要先使用vela def get configmap查看对应trait的cue定义
image
然后我才能知道如何填充map[string]interface{},这个过程我需要重复多次，因为在之前的k8s原生对象中有多个trait（configmap, label等）,这样看就已经背离我最开始的期望了

**Describe the solution you'd like**，
有没有一种方法可以从cue中生成对应的go struct，这样我们只要导入对应的struct，就可以获取到具体的参数，而不是都是interface{}
如下：
```
properties: {
 name: string
age?: int
} 
转换为 
type WebServiceProperties struct {
  Name string `required`
 Age int 
}
```

**Describe alternatives you've considered**
对于现在而言，我已经在考虑模板语法，去复制多个yaml了

**Additional context**
vela的确有很强的功能，但是我们是不是应该先保障到最核心的功能，比如cue的渲染能力和app这样一个核心对象，我觉得在转换资源上我应该不是第一个遇到该问题的，社区有好的建议我们也可以一起讨论

### PR DESCRIPTION
**Pull request by chivalryq**
Signed-off-by: qiaozp <chivalry.pp@gmail.com>

### Description of your changes

Add `vela def gen-api` command to generate Go struct for X-definition, example: 

```shell
$ vela def go-gen vela-templates/definitions/internal/trait/resource.cue
package main

// Requests Specify the resources in requests
type Requests struct {
        CPU    number `json:"cpu"`
        Memory string `json:"memory"`
}

// Limits Specify the resources in limits
type Limits struct {
        CPU    number `json:"cpu"`
        Memory string `json:"memory"`
}

// Parameter -
type Parameter struct {
        CPU      number   `json:"cpu"`
        Memory   string   `json:"memory"`
        Requests Requests `json:"requests"`
        Limits   Limits   `json:"limits"`
}

```

Partially fixes #3606 

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

1. A lot tests with different types of parameter in go_gen_test.go
2. Test almost every X-definition in `vela-templates/definitions/`. Some of them are skipped now because it need more investigation about how to convert them to Go struct.

### CODE REVIEW

**Comment by Somefive** [comment]
Great work! 

BTW, how could this command be used in the api library? Is there any script that can leverage it to generate the go code lib for all definitions?

**Comment by chivalryq** [comment]
> Great work!
> 
> BTW, how could this command be used in the api library? Is there any script that can leverage it to generate the go code lib for all definitions?

This will be done in next PR. I suggest we add one CI to sync all go code to another repo.

**Comment by zzxwill** [comment]
The tool is that the community and I have waited for long.

How about periodically generating the API structs and merging them into [oam-dev/](https://github.com/oam-dev/kubevela-core-api). Platform builders will use the Go APIs.

File: pkg/definition/go_gen.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+func genField(param StructParameter, buffer *bytes.Buffer) {
+	fieldName := dm.FieldName(param.Name)
+	if param.Type == cue.StructKind { // only struct kind will be separated struct
+		// cue struct  can be Go map or struct
+		if strings.HasPrefix(param.GoType, "map[string]") {
+			fmt.Fprintf(buffer, "type %s %s", fieldName, param.GoType)
+		} else {
+			fmt.Fprintf(buffer, "type %s struct {\n", fieldName)
+			for _, f := range param.Fields {
+				fmt.Fprintf(buffer, "    %s %s `json:\"%s\"`\n", dm.FieldName(f.Name), f.GoType, f.Name)
```
Comment (barnettZQG) [inline pkg/definition/go_gen.go:216]: Add the `omitempty` JSON tag for the field that is not required.

Comment (chivalryq) [reply pkg/definition/go_gen.go]: done that

**Review by Somefive — APPROVED** [review]
Great Job! LGTM

**Review by wonderflow — APPROVED** [review]

**Pull request merged**